src/app/api/tasks/tasks.service.ts

```diff
@@ -529,24 +541,29 @@ export class TasksService extends BaseService {
 
     // Case 3
     // --- If task was previously in another state, and is moved to a 'completed' type WorkflowState by IU
+    let shouldCreateNotification = true
     if (
       prevTask?.workflowState?.type !== StateType.completed &&
       updatedTask?.workflowState?.type === StateType.completed &&
       updatedTask.assigneeId
     ) {
       // Don't send task notifications if the IU created the task themselves
       if (updatedTask.createdById === this.user.internalUserId) {
-        return
+        shouldCreateNotification = false
       }
 
       if (updatedTask.assigneeType === AssigneeType.internalUser) {
-        await notificationService.create(NotificationTaskActions.CompletedByIU, updatedTask, { email: true })
+        shouldCreateNotification &&
+          (await notificationService.create(NotificationTaskActions.CompletedByIU, updatedTask, { email: true }))
         // TODO: Clean code and handle notification center notification deletions here instead
       } else if (updatedTask.assigneeType === AssigneeType.company) {
         // Don't do this in parallel since this can cause rate-limits, each of them has their own bottlenecks for avoiding ratelimits
-        await notificationService.create(NotificationTaskActions.CompletedForCompanyByIU, updatedTask, { email: true })
+        shouldCreateNotification &&
+          (await notificationService.create(NotificationTaskActions.CompletedForCompanyByIU, updatedTask, { email: true }))
         await notificationService.markAsReadForAllRecipients(updatedTask)
       } else if (updatedTask.assigneeType === AssigneeType.client) {
+        shouldCreateNotification &&
```

Comment:
so if the IU is the creator of the task,
and assignee is the client, then we dont send NotificationTaskActions.CompletedByIU?

Reply:
There are two ways that can happen in this context.

1. IU assigns the task to client, but that IU / another IU completes the task. If the task is created by the IU themselves, it is caught by `shouldCreateNotification` flag check above. If the task is marked as done by some other IU, the notification says that this client task has been completed by IU, hence `NotificationTaskActions.CompletedByIU`

2. This client, or another company client completes this task. This is handled by another service to keep concerns separate and isn't affected here. I did verify this behavior again in this PR, and it is working as intended.

Reply:
We aren't sending new notifications to client when task is completed. Just removing it from his task notifications count.
See the usage of `markClientNotificationAsRead` here
![image](https://github.com/user-attachments/assets/28e49122-9f55-421e-84dd-9f924b440ea2)